File: src/turtlesim_catch_them_all/turtlesim_catch_them_all/turtle_spawner.py
#!/usr/bin/env python3 
import rclpy 
from rclpy.node import Node 
import random
from functools import partial

from turtlesim.srv import Spawn
from turtlesim.srv import Kill
from my_robot_interfaces.msg import Turtle
from my_robot_interfaces.srv import CatchTurtle

class TurtleSpawnerNode(Node):
    def __init__(self):
        super().__init__("turtle_spawner")
        self.declare_parameter("spawn_frequency", 1.0)
        self.spawn_frequency_ = self.get_parameter("spawn_frequency").value
        self.declare_parameter("object_name", "turtle_")
        self.object_name_ = self.get_parameter("object_name").value
        self.turtles_ = []
        self.turtles_x_ = []
        self.turtles_y_ = []
        self.counter_ = 0
        self.spawner_client_ = self.create_client(Spawn, "spawn")
        self.killer_client_ = self.create_client(Kill, "kill")
        self.spawner_timer_ = self.create_timer(self.spawn_frequency_, self.call_spwaner)
        self.alive_turtles_publisher_ = self.create_publisher(Turtle, "alive_turtles", 10)
        self.alive_turtles_timer_ = self.create_timer(0.3, self.publisher_alive_turtles)
        self.catch_turtle_server_ = self.create_service(CatchTurtle, "catch_turtle", self.callback_catch_turtle)

    def call_spwaner(self):
        spawner_request = Spawn.Request()
        spawner_request.x = round(random.uniform(0,11),1)
        spawner_request.y = round(random.uniform(0,11),1)
        self.counter_ += 1
        spawner_request.name = self.object_name_ + str(self.counter_)
        # self.get_logger().info(spawner_request.name)
        # self.spawn_x_ = spawner_request.x
        # self.spawn_y_ = spawner_request.y
        future = self.spawner_client_.call_async(spawner_request)
        future.add_done_callback(partial(self.callback_spwaner, spawner_request = spawner_request)) # Partial can bring parameter to callback

    def callback_spwaner(self, future, spawner_request: Spawn.Request):
        spawner_response = future.result()
        self.turtles_x_.append(spawner_request.x)
        self.turtles_y_.append(spawner_request.y)
        self.turtles_.append(spawner_response.name)
        self.get_logger().info("We have " + str(self.turtles_))
            
    def publisher_alive_turtles(self):
        msg = Turtle()
        msg.name = self.turtles_
        msg.x = self.turtles_x_
        msg.y = self.turtles_y_
        self.alive_turtles_publisher_.publish(msg)
            
    def callback_catch_turtle(self, request: CatchTurtle.Request, response: CatchTurtle.Response):
        self.turtle_name_ = request.name
        target_turtle_index = self.turtles_.index(self.turtle_name_)
        killer_request = Kill.Request()
        killer_request.name = self.turtle_name_
        future = self.killer_client_.call_async(killer_request)
        future.add_done_callback(partial(self.callback_killer, index = target_turtle_index))        
        response.success = True
        return response
     
    def callback_killer(self, future, index):
        del self.turtles_[index]
        del self.turtles_x_[index]
        del self.turtles_y_[index]
        
        # self.turtles_.remove(self.turtles_[self.target_turtle_index_])
        # self.turtles_x_.remove(self.turtles_x_[self.target_turtle_index_])
        # self.turtles_y_.remove(self.turtles_y_[self.target_turtle_index_])
        # self.get_logger().info(self.turtle_name_ + " has been killed")

        


def main(args=None): 
    rclpy.init(args=args) 
    node = TurtleSpawnerNode()  
    rclpy.spin(node)    
    rclpy.shutdown() 

if __name__ == "__main__":  
    main()

File: src/Will_cpp_pkg/src/turtle_spawner.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtlesim/srv/spawn.hpp"
#include "turtlesim/srv/kill.hpp"
#include <random>
#include "my_robot_interfaces/msg/turtle2.hpp"
#include "my_robot_interfaces/msg/turtle2_array.hpp"
#include "my_robot_interfaces/srv/catch_turtle.hpp"

using namespace std::chrono;
using namespace std::placeholders;

// struct Turtle
// {
//     float x;
//     float y;
//     std::string name;
// };

class TurtleSpawnerNode : public rclcpp::Node
{
public:
    TurtleSpawnerNode() : Node("turtle_spawner"), counter_(0)
    {
        this->declare_parameter("prefix", "frog");
        prefix_ = this->get_parameter("prefix").as_string();
        this->declare_parameter("spawn_rate", 2);
        int spawn_frequency = this->get_parameter("spawn_rate").as_int();
        spawn_ = this->create_client<turtlesim::srv::Spawn>("spawn", 10);
        kill_ = this->create_client<turtlesim::srv::Kill>("kill",10);
        timer_ = this->create_wall_timer(seconds(spawn_frequency), std::bind(&TurtleSpawnerNode::spwanTurtle, this));
        turtle_pub_ = this->create_publisher<my_robot_interfaces::msg::Turtle2Array>("alive_turtle", 10);
        catch_it_ = this->create_service<my_robot_interfaces::srv::CatchTurtle>("catch_turtle", std::bind(&TurtleSpawnerNode::killTurtle, this, _1, _2));

    }
private:
    void spwanTurtle()
    {
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<double> dist_xy(0.0, 11.0);
        std::uniform_real_distribution<double> dist_theta(- M_PI, M_PI);
        
        auto request = std::make_shared<turtlesim::srv::Spawn::Request>();
        request->x = dist_xy(gen);
        request->y = dist_xy(gen);
        request->theta = dist_theta(gen);
        request->name = prefix_ + "_" + std::to_string(++counter_);
        spawn_->async_send_request(request, 
            [this, request](rclcpp::Client<turtlesim::srv::Spawn>::SharedFuture future)
            {
            my_robot_interfaces::msg::Turtle2 turtle;
            auto response = future.get();
            turtle.x = request->x;
            turtle.y = request->y;
            turtle.name = response->name;
            turtles.push_back(turtle);
            auto msg = my_robot_interfaces::msg::Turtle2Array();
            msg.turtles = turtles;
            turtle_pub_->publish(msg);
            });
    }

    void killTurtle(const my_robot_interfaces::srv::CatchTurtle::Request::SharedPtr request,
                          my_robot_interfaces::srv::CatchTurtle::Response::SharedPtr response)
    {
        int index = -1;
        for (size_t i = 0; i < turtles.size(); i++)
        {
            if (turtles[i].name == request->name)
                {
                    index = i;
                    break;
                }
        }
        if (index == -1)
            {
                RCLCPP_INFO(this->get_logger(),"index = -1 ");
            }
        auto request2 = std::make_shared<turtlesim::srv::Kill::Request>();
        request2->name = turtles[index].name;
        kill_->async_send_request(request2,
            [this, index](rclcpp::Client<turtlesim::srv::Kill>::SharedFuture future)
            {
                auto response2 = future.get();
                RCLCPP_INFO(this->get_logger(),"%s has been killed !", turtles[index].name.c_str());
                turtles.erase(turtles.begin() + index);
                auto msg = my_robot_interfaces::msg::Turtle2Array();
                msg.turtles = turtles;
                turtle_pub_->publish(msg);
            });
        response->success = true;
    }


    rclcpp::Client<turtlesim::srv::Spawn>::SharedPtr spawn_;
    rclcpp::Client<turtlesim::srv::Kill>::SharedPtr kill_;
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::TimerBase::SharedPtr timer_test_;
    rclcpp::Publisher<my_robot_interfaces::msg::Turtle2Array>::SharedPtr turtle_pub_;
    std::vector<my_robot_interfaces::msg::Turtle2> turtles;
    rclcpp::Service<my_robot_interfaces::srv::CatchTurtle>::SharedPtr catch_it_;
    std::string prefix_;
    int counter_;
    
    
        
};



int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<TurtleSpawnerNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}